rfcs/0112-demote-darwin-support-tier.md

```diff
@@ -0,0 +1,91 @@
+---
+feature: demote_darwin_support_tier
+start-date: 2021-10-27
+author: piegames
+co-authors: many
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Move x86_64-darwin from a Tier 2 platform to Tier 3, as described in [RFC 0046](https://github.com/NixOS/rfcs/blob/master/rfcs/0046-platform-support-tiers.md).
+
+# Motivation
+[motivation]: #motivation
+
+Darwin is constantly adding additional maintainer burden. Especially, it does not live up to the requirements:
```

Comment:
For the 21.05 release, I decided to move forward when darwin rebuilds were ~6k in staging-next jobset, as it was mostly haskell packages. I'm not sure how many nix-darwin-haskell users we have, but having to lag behind unstable for a few days seems reasonable when it causes risk to release dates.

Reply:
> Maybe another data point from my experience maintaining `haskellPackages`:
> 
> * Hydra has been unreliable for getting darwin builds done. This has been a major pain point and seriously limited our ability to finish `haskell-updates` rotations in a sensible time frame. As a result we have [decided to ignore](https://github.com/NixOS/nixpkgs/pull/126903) queued darwin builds when merging `haskell-updates`. This is, however, only possible because nothing in `haskellPackages` relates to a darwin channel blocker, so this “just” slows down trunk evaluations building. For `staging-next` darwin has been _the_ problem slowing down rotations.

I've noted in other comments what's going on with Hydra/ofborg and darwin workers, could we aim to have one thread for that?

> * Darwin is not just another platform: Maybe a bit obvious, but I want to point out that portability to darwin is much worse than for other Linux architectures: Upstream support is usually worse, it's harder to get help from upstream and the actual portability issues are often non-trivial to solve.

The reason we care so much about macOS is that it's so widely used among the developers - depending on what statistic you take a look at, it's somewhere [around 45% of market share](https://www.statista.com/statistics/869211/worldwide-software-development-operating-system/#:~:text=Software%20development%20operating%20system%20distribution%20globally%202018%20to%202021&text=Most%20software%20developers%20worldwide%20report,percent%20of%20developers%20preferring%20Linux.), barely a few percent after Linux.

It's a lot harder to support, but that shouldn't be the metric of itself, but rather for how long it's allowed to Tier 1 platforms.

Reply:
> I've noted in other comments what's going on with Hydra/ofborg and darwin workers, could we aim to have one thread for that?

I'll reply there, sure. However my comment was more about how this impacts development and maintenance in nixpkgs.

> The reason we care so much about macOS is that it's so widely used among the developers - depending on what statistic you take a look at, it's somewhere around 45% of market share, barely a few percent after Linux.

I find it quite instructive that this comment reads like a sales pitch: While I do subscribe to nixpkgs' emerging “build all the things” philosophy and would love to make nixpkgs available to every conceivable user, it is not my *primary* concern. The primary concern needs to be the ability to maintain whatever we support in a sustainable fashion. In my experience this is not the case and I don't want to burden contributors that don't use darwin with the extra work involved, like having to invest a full day into fixing a darwin regression introduced by a change trivial to implement for Linux.

Also I would like to ask you who the “we” you are referring to is. I think this RFC demonstrate how unenthusiastic some core contributors about supporting darwin as a Tier 2 platform.

> It's a lot harder to support, but that shouldn't be the metric of itself, but rather for how long it's allowed to Tier 1 platforms.

Not sure I quite understand what you mean by the second part of this sentence, but I do think it is the most important metric as outlined above. nixpkgs is an open source project, so everything it is should be dictated by what its contributors want and can pull off.

Reply:
> Darwin is not just another platform: Maybe a bit obvious, but I want to point out that portability to darwin is much worse than for other Linux architectures: Upstream support is usually worse, it's harder to get help from upstream and the actual portability issues are often non-trivial to solve.

Worth also noting that our darwin platform is unusual even to upstreams which support Linux + macOS - they typically expect XCode + BSD coreutils + Apple Clang, while we use "fake" xcbuild, GNU coreutils, and mainline LLVM Clang. So the portability issues are nontrivial even when macOS is ostensibly supported upstream.

Reply:
At the risk of further diverging this tangent...

I think linux users *do* actually gain something from nixpkgs' darwin support which we tend to ignore - much better clang support almost across the board.

Reply:
> > I've noted in other comments what's going on with Hydra/ofborg and darwin workers, could we aim to have one thread for that?
> 
> I'll reply there, sure. However my comment was more about how this impacts development and maintenance in nixpkgs.
> 
> > The reason we care so much about macOS is that it's so widely used among the developers - depending on what statistic you take a look at, it's somewhere around 45% of market share, barely a few percent after Linux.
> 
> I find it quite instructive that this comment reads like a sales pitch: While I do subscribe to nixpkgs' emerging “build all the things” philosophy and would love to make nixpkgs available to every conceivable user, it is not my _primary_ concern. The primary concern needs to be the ability to maintain whatever we support in a sustainable fashion. In my experience this is not the case and I don't want to burden contributors that don't use darwin with the extra work involved, like having to invest a full day into fixing a darwin regression introduced by a change trivial to implement for Linux.

I don't think anyone should be forced into fixing darwin breakages. We should organize better to have macOS contributors around that help out, but as part of tier 2 darwin has a limited time to response (which is unfortunately unspecified).

My comment wasn't meant as a sales pitch, but rather to build empathy to our users and why it's important that we try to fix things rather than make things worse (which is what this tier demotion would do).

> 
> Also I would like to ask you who the “we” you are referring to is. I think this RFC demonstrate how unenthusiastic some core contributors about supporting darwin as a Tier 2 platform.

As I've said, noone is forced to contribute darwin support. What is happening is a misunderstanding that someone is expected to fix their change to build on darwin. Tier 2 allows merging changes if there are no darwin fixes after a certain time (that time is yet to be determined, but that's a topic for the tiers RFC).

> 
> > It's a lot harder to support, but that shouldn't be the metric of itself, but rather for how long it's allowed to Tier 1 platforms.
> 
> Not sure I quite understand what you mean by the second part of this sentence, but I do think it is the most important metric as outlined above. nixpkgs is an open source project, so everything it is should be dictated by what its contributors want and can pull off.

100%.